lib/jsonapi_parameters/translator.rb

```diff
@@ -50,7 +50,10 @@ def jsonapi_main_body
   end
 
   def jsonapi_unsafe_params
-    @jsonapi_unsafe_params ||= @jsonapi_unsafe_hash.dig(:data, :attributes) || {}
+    @jsonapi_unsafe_params ||= @jsonapi_unsafe_hash
+                               .dig(:data, :attributes)
+                               .merge(id: @jsonapi_unsafe_hash.dig(:data, :id))
+                               .compact || {}
```

Comment:
I am concerned about this part. 

* I would prefer not to call `.merge` if there is no reason to (e.g. there is no `:id` key within `@jsonapi_unsafe_hash.dig(:data)`),
* Calling `.compact` on `data.attributes` will remove all key-val pairs where a value is set to `nil`. I am not sure if this is what we want; please correct me if I am wrong, but I think the specification allows the client to explicitly pass `null` values, and these should be accepted.

Maybe it is worth going for something like this:

```ruby
  def jsonapi_unsafe_params
    @jsonapi_unsafe_params ||= (@jsonapi_unsafe_hash.dig(:data, :attributes) || {}).tap do |param|
      param[:id] = @jsonapi_unsafe_hash.dig(:data, :id) if @jsonapi_unsafe_hash[:data].key?(:id)
    end
  end
```

What do you think?

Reply:
Good thinking! 
Makes sense not to remove data that might be explicitly being set as null. and I like the solution with `.tap`!

Reply:
My only concern with this is that there will need to be explicit declaration that this change is being made. Take the scenario: 

`PATCH /accounts/1234`

```json
{
  data: {
    type: "account",
    id: nil,
    attributes: {
      name: "Accountant Accounting 2019"
    }

  }
}
```
This _will_ remove the primary key from the record. Doing this locally, made some tests fail as I wasn't explicitly setting ID in all of my JSON objects. My tests catching this is good, but not everyone will have them. 

I see two solutions: 
1. I can update the documentation to reflect that this change is potentially breaking,
2. We can "soft" introduce this by asserting presence of an ID value, rather than an ID key. I cannot for the life of me see the need for a record with no primary key, and even updating it is a highly rare case.

Reply:
Something like 
```ruby
 def jsonapi_unsafe_params
    @jsonapi_unsafe_params ||= (@jsonapi_unsafe_hash.dig(:data, :attributes) || {}).tap do |param|
      param[:id] = @jsonapi_unsafe_hash.dig(:data, :id) if @jsonapi_unsafe_hash.dig(:data, :id).present?
    end
  end
```

Reply:
Hmm... as I understand, you're worried about the fact that users may pass `id` that has a `nil` value? It may indeed result in nullifying the record (assuming someone does not filter the parameter out). ActiveRecord itself probably won't let that happen though. 

Neither can I imagine a scenario where such feature may be required, but if there's one, why shouldn't we let users do it?
This library should focus most on just translating the parameter structure, and doing safe parsing / filtering should be limited in my opinion. 

I'm always in favour of giving users the freedom, so maybe it would be a good idea to set a flag, such as:

`params.from_jsonapi(allow_nil_id: true)`, which would let `nil` IDs pass; without setting it explicitly, we would go for your proposed code, `param[:id] = @jsonapi_unsafe_hash.dig(:data, :id) if @jsonapi_unsafe_hash.dig(:data, :id)`.

What do you think? 

Also, in regards to the latest code paste, lets avoid calling `.dig` with same parameters twice -  lets just assign it to variable and then reuse it; its a micro optimisation, but I think cheap enough to implement :), so something like this:

```ruby
 def jsonapi_unsafe_params
    @jsonapi_unsafe_params ||= (@jsonapi_unsafe_hash.dig(:data, :attributes) || {}).tap do |param|
      id = @jsonapi_unsafe_hash.dig(:data, :id)

      param[:id] = id if id.present?
    end
  end
```

Reply:
You're right on all accounts. I made the argument that it shouldn't be the library's job to safe parse out an ID and here I am doing the same thing. 

For the record: 
```ruby
[1] pry(main)> Contact.last.id
"fe5f4f61-23de-4889-8705-4cab541d1166"
[2] pry(main)> Contact.last.update(id: nil)
   ...
   (0.1ms)  ROLLBACK
ActiveRecord::NotNullViolation: PG::NotNullViolation: ERROR:  null value in column "id" violates not-null constraint
```

You're right, it won't let you nullify a primary key, You can however re-write a primary key.

```ruby
[3] pry(main)> Contact.last.update(id: SecureRandom.uuid)
true
```

With our above logic of "if you can do it, it's not up to us to say you can't", we should let someone update their primary key _if they want to_.

I've gone with the implementation of "if the ID value exists lets `update/create` otherwise let's "leave it alone /server generate it". The reason for the _value_ presence is `nil` on `update` will generate an error.

Should we introduce the ability to generate an error for nil on update?

Reply:
You're right on all accounts. I made the argument that it shouldn't be the library's job to safe parse out an ID and here I am doing the same thing. 

For the record: 
```ruby
irb(main):001:0> Post.create(title: "hello world")
   (0.2ms)  BEGIN
  Post Create (5.8ms)  INSERT INTO "posts" ("title", "created_at", "updated_at") VALUES ($1, $2, $3) RETURNING "id"  [["title", "hello world"], ["created_at", "2019-10-27 22:49:36.837796"], ["updated_at", "2019-10-27 22:49:36.837796"]]
   (4.1ms)  COMMIT
=> #<Post id: 1, title: "hello world", created_at: "2019-10-27 22:49:36", updated_at: "2019-10-27 22:49:36">
irb(main):002:0> Post.last.update(id: 123)
  Post Load (0.8ms)  SELECT "posts".* FROM "posts" ORDER BY "posts"."id" DESC LIMIT $1  [["LIMIT", 1]]
   (0.2ms)  BEGIN
  Post Update (0.4ms)  UPDATE "posts" SET "id" = $1, "updated_at" = $2 WHERE "posts"."id" = $3  [["id", 123], ["updated_at", "2019-10-27 22:49:46.154769"], ["id", 1]]
   (1.9ms)  COMMIT
=> true
irb(main):003:0> Post.last.update(id: nil)
  Post Load (0.7ms)  SELECT "posts".* FROM "posts" ORDER BY "posts"."id" DESC LIMIT $1  [["LIMIT", 1]]
   (0.3ms)  BEGIN
  Post Update (2.5ms)  UPDATE "posts" SET "id" = $1, "updated_at" = $2 WHERE "posts"."id" = $3  [["id", nil], ["updated_at", "2019-10-27 22:49:50.423845"], ["id", 123]]
   (0.2ms)  ROLLBACK
Traceback (most recent call last):
        1: from (irb):3
ActiveRecord::NotNullViolation (PG::NotNullViolation: ERROR:  null value in column "id" violates not-null constraint)
DETAIL:  Failing row contains (null, hello world, 2019-10-27 22:49:36.837796, 2019-10-27 22:49:50.423845).
```
Rails default won't let you nullify a primary key, you can however re-write a primary key.

With that logic, and the above logic (it's not up to us to decide what users do once the library's jurisdiction ends), it makes the most sense that if a `Primary ID Key` is present, we should be adding it to the attributes hash. It's there, the User is attempting to CRUD it, it's the library's job to facilitate that. I'll update the code accordingly.

EDIT: 
I deleted the original comment and added this one because I realised I was just typing my thoughts out and editing the comment. Not wanted my entire stream of consciousness here just decided to go with the final product.